op-chain-ops/crossdomain/legacy_withdrawal.go

```diff
@@ -0,0 +1,132 @@
+package crossdomain
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum-optimism/optimism/op-bindings/predeploys"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+// LegacyWithdrawal represents a pre bedrock upgrade withdrawal.
+type LegacyWithdrawal struct {
+	Target *common.Address
+	Sender *common.Address
+	Data   []byte
+	Nonce  *big.Int
+}
+
+var _ WithdrawalMessage = (*LegacyWithdrawal)(nil)
+
+// NewLegacyWithdrawal will construct a LegacyWithdrawal
+func NewLegacyWithdrawal(target, sender *common.Address, data []byte, nonce *big.Int) *LegacyWithdrawal {
+	return &LegacyWithdrawal{
+		Target: target,
+		Sender: sender,
+		Data:   data,
+		Nonce:  nonce,
+	}
+}
+
+// Encode will serialze the Withdrawal in the legacy format so that it
+// is suitable for hashing. This assumes that the message is being withdrawn
+// through the standard optimism cross domain messaging system by hashing in
+// the L2CrossDomainMessenger address.
+func (w *LegacyWithdrawal) Encode() ([]byte, error) {
+	enc, err := EncodeCrossDomainMessageV0(w.Target, w.Sender, w.Data, w.Nonce)
+	if err != nil {
+		return nil, err
+	}
+
+	out := make([]byte, len(enc)+len(predeploys.L2CrossDomainMessengerAddr.Bytes()))
+	copy(out, enc)
+	copy(out[len(enc):], predeploys.L2CrossDomainMessengerAddr.Bytes())
+	return out, nil
+}
+
+// Decode will decode a serialized LegacyWithdrawal
+func (w *LegacyWithdrawal) Decode(data []byte) error {
+	if len(data) < 4 {
```

Comment:
The length check should be that is greater than 4 + len(address). This is because on line 71 we slice from `4` to `len(data) - len(address)`. We have only guaranteed that `len(data) >= len(address) >= 4`. That means that if `len(data) == len(address)`, slicing from `4:0` on line 71 would panic.

Reply:
I've addressed this in https://github.com/ethereum-optimism/optimism/commit/dd274000bcbdc3325f11e84918d6381ced10786c